File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/WorkPostsController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.db.Datastore
import net.animeta.backend.db.query
import net.animeta.backend.dto.PostDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.QHistory.history
import net.animeta.backend.model.StatusType
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.WorkRepository
import net.animeta.backend.serializer.PostSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/works/{id:[0-9]+}/posts")
class WorkPostsController(val workRepository: WorkRepository,
                          val historyRepository: HistoryRepository,
                          val datastore: Datastore,
                          val postSerializer: PostSerializer) {
    data class GetWithCountsResponse(val data: List<PostDTO>, val userCount: Int?, val suspendedUserCount: Int?)

    @GetMapping
    fun get(@PathVariable id: Int,
            @RequestParam("before_id", required = false) beforeId: Int?,
            @RequestParam(required = false) episode: String?,
            @RequestParam(defaultValue = "32") count: Int,
            @RequestParam(defaultValue = "false") withCounts: Boolean,
            @RequestParam(required = false) options: PostSerializer.Options?): Any {
        val work = workRepository.findById(id).orElseThrow { ApiException.notFound() }
        var query = history.query.selectRelated(history.user)
                .filter(history.workId.eq(work.id!!), history.comment.ne(""))
                .orderBy(history.id.desc())
        if (beforeId != null) {
            query = query.filter(history.id.lt(beforeId))
        }
        if (episode != null) {
            query = query.filter(history.status.eq(episode))
        }
        val limit = minOf(count, 128)
        val posts = datastore.query(query.limit(limit))
                .map { postSerializer.serialize(it, options ?: PostSerializer.legacyOptions(includeUser = true)) }
        if (!withCounts) {
            return posts
        }

        val (userCount, suspendedUserCount) = if (episode != null) {
            val counts = historyRepository.countDistinctUsersByStatusType(work.id!!, episode)
            Pair(
                counts.sumBy { it.second.toInt() },
                counts.find { it.first == StatusType.SUSPENDED }?.second?.toInt() ?: 0
            )
        } else {
            Pair(null, null) // TODO
        }
        return GetWithCountsResponse(
            data = posts,
            userCount = userCount,
            suspendedUserCount = suspendedUserCount
        )
    }
}

File: backend-ts/src/controllers/work_posts.controller.ts
import { Controller, DefaultValuePipe, Get, Param, ParseBoolPipe, ParseIntPipe, Query } from "@nestjs/common";
import { PostDTO } from 'shared/types';
import { ApiException } from "./exceptions";
import { LessThan, Not, Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";
import { PostSerializer, PostSerializerOptions } from "src/serializers/post.serializer";
import { History } from 'src/entities/history.entity'
import { ParseIntOrNullPipe } from "src/utils/parse_int_or_null.pipe";
import { Work } from "src/entities/work.entity";
import { StatusType } from "src/entities/status_type";

export type GetWorkPostsWithCountsResponse = {
  data: PostDTO[];
  userCount: number | null;
  suspendedUserCount: number | null;
}

@Controller('/api/v4/works/:id/posts')
export class WorkPostsController {
  constructor(
    @InjectRepository(Work) private workRepository: Repository<Work>,
    @InjectRepository(History) private historyRepository: Repository<History>,
    private postSerializer: PostSerializer,
  ) {}

  @Get()
  async get(
    @Param('id', new ParseIntPipe()) id: number,
    @Query('before_id', new ParseIntOrNullPipe()) beforeId: number | null,
    @Query('episode') episode: string | null,
    @Query('count', new ParseIntOrNullPipe(), new DefaultValuePipe(32)) count: number,
    @Query('withCounts', new DefaultValuePipe('false'), new ParseBoolPipe()) withCounts: boolean,
    @Query('options') optionsJson: string | null,
  ): Promise<PostDTO[] | GetWorkPostsWithCountsResponse> {
    const work = await this.workRepository.findOne(id)
    if (!work) throw ApiException.notFound()
    const posts = await this.historyRepository.find({
      relations: ['user'],
      where: {
        work_id: work.id,
        comment: Not(''),
        ...beforeId ? { id: LessThan(beforeId) } : {},
        ...episode ? { status: episode } : {},
      },
      order: { id: 'DESC' },
      take: Math.min(count, 128),
    })
    const options: PostSerializerOptions =
      optionsJson ? JSON.parse(optionsJson) : PostSerializer.legacyOptions({
        includeUser: true,
      })
    const data = await Promise.all(posts.map(it => this.postSerializer.serialize(it, options)))
    if (!withCounts)
      return data
    
    let userCount: number | null = null
    let suspendedUserCount: number | null = null
    if (episode) {
      const counts = (await this.historyRepository.createQueryBuilder()
        .select('status_type')
        .addSelect('COUNT(DISTINCT user_id)', 'count')
        .where('work_id = :workId AND status = :status', { workId: work.id, status: episode })
        .groupBy('status_type')
        .getRawMany())
        .map(it => ({
          statusType: it.status_type as StatusType,
          count: Number(it.count),
        }))
      userCount = counts.reduce((acc, it) => acc + it.count, 0)
      suspendedUserCount = counts.filter(it => it.statusType === StatusType.SUSPENDED)
        .reduce((acc, it) => acc + it.count, 0)
    } else {
      // TODO
    }
    return {
      data,
      userCount,
      suspendedUserCount,
    }
  }
}
